File: src/main/java/io/saso/dash/redis/tables/DashRedisConnections.java
package io.saso.dash.redis.tables;

import com.google.inject.Inject;
import com.google.inject.Singleton;
import io.saso.dash.redis.Redis;
import io.saso.dash.redis.RedisDatabase;
import org.apache.logging.log4j.LogManager;
import redis.clients.jedis.Jedis;

@Singleton
public class DashRedisConnections implements RedisConnections
{
    private static final int MAX_CONCURRENT_CONNECTIONS_PER_USER = 3;

    private final Redis redis;

    @Inject
    public DashRedisConnections(Redis redis)
    {
        this.redis = redis;

        clear();
    }

    @Override
    public boolean addIfAllowed(int userId)
    {
        final String key = String.valueOf(userId);

        try (Jedis conn = redis.getConnection(
                RedisDatabase.CONCURRENT_CONNECTIONS)) {
            final long connCount = conn.incr(key);

            LogManager.getLogger().debug("{} concurrent connections for user " +
                    "{}", connCount, userId);

            if (connCount > MAX_CONCURRENT_CONNECTIONS_PER_USER) {
                conn.decr(key);
                return false;
            }
        }

        return true;
    }

    @Override
    public void remove(int userId)
    {
        try (Jedis conn = redis.getConnection(
                RedisDatabase.CONCURRENT_CONNECTIONS)) {
            conn.decr(String.valueOf(userId));
        }
    }

    private void clear()
    {
        try (Jedis conn = redis.getConnection(
                RedisDatabase.CONCURRENT_CONNECTIONS)) {
            conn.flushDB();
        }
    }
}

File: src/main/kotlin/io/saso/dash/redis/databases/DashRedisConnections.kt
package io.saso.dash.redis.databases

import com.google.inject.Inject
import com.google.inject.Singleton
import io.saso.dash.config.Config
import io.saso.dash.redis.Redis
import io.saso.dash.redis.RedisDatabase
import io.saso.dash.util.key
import io.saso.dash.util.logger

@Singleton
public class DashRedisConnections
@Inject
constructor(private val redis: Redis, private val config: Config) :
        RedisConnections
{
    private val maxConcurrentConnectionsPerUser =
            config.get("max-connections-per-user", 3)

    override fun addIfAllowed(userId: Int): Boolean {
        val key = key("user", userId)

        redis.getConnection(RedisDatabase.CONCURRENT_CONNECTIONS) use {
            val connCount = it incr key

            if (connCount > maxConcurrentConnectionsPerUser) {
                it decr key
                return false
            }

            logger(this@DashRedisConnections) debug
                    "Concurrent connnections for user $userId: $connCount"
        }

        return true
    }

    override fun remove(userId: Int) {
        redis getConnection RedisDatabase.CONCURRENT_CONNECTIONS use {
            it decr key("user", userId)
        }
    }
}
